File: find-minimum-in-rotated-sorted-array.py
# // Time Complexity : O(Log N)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : YES
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
class Solution:
    def findMin(self, nums: List[int]) -> int:
        start , end = 0, len(nums) - 1 
        curr_min = 5000 #Max Allowed Number in the question
        
        while start  <  end :
            mid = (start + end ) // 2
            curr_min = min(curr_min,nums[mid])
            
            # right has the min 
            if nums[mid] > nums[end]:
                start = mid + 1
                
            # left has the  min 
            else:
                end = mid - 1 
                
        return min(curr_min,nums[start])

File: find-minimum-in-rotated-sorted-array.java
// Time Complexity : o(logn)
// Space Complexity :o(1)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : NO

//we are performing the binary search in the unsorted part of the array, 
//and finding the minimum element

class Solution {
    public int findMin(int[] nums) {
        // base case
        if (nums.length == 1)
            return nums[0];
        // logic
        int n = nums.length;
        int l = 0, r = n - 1;
        while (l <= r) {
            if (nums[l] <= nums[r])
                return nums[l];
            int mid = l + (r - l) / 2;
            if ((mid == 0 || nums[mid] < nums[mid - 1]) && (mid == n - 1 || nums[mid] < nums[mid + 1]))
                return nums[mid];
            else if (nums[l] <= nums[mid])
                l = mid + 1;
            else
                r = mid - 1;
        }
        return 0;
    }
}